### PR TITLE
Support Odroid C4 SBC

### PR DESCRIPTION
**Pull request by ubergeek801**
This PR adds support for the Odroid C4 SBC, based on the existing configuration for the Odroid N2 (which is very similar). I have tested the build on both SD and eMMC media, verifying that the OS boots, configures itself, permits a user account to be created, etc.

Items that may require particular attention:
- In relevant documentation, I have marked C4 support as "experimental"; I don't know whether this is appropriate.
- In the `meta` configuration, the string `odroid-n2` is currently used for the `BOARD_ID` and `SUPERVISOR_MACHINE`, as it appears that the string has to match something at https://version.home-assistant.io/stable.json. Presumably `odroid-c4` could be added here.
- Similar to the Odroid N2, the blobs `acs.bin` and `bl301.bin` are provided from a source outside this repo. For now, `hardkernel-boot.mk` obtains them from Neil Armstrong's repo at https://gitlab.com/superna9999/amlogic-boot. Presumably the "blessed" versions (wherever they come from) would be added to `hassos-blobs` as with the N2, or maybe the C4's blobs are actually the same as the N2's?

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @ubergeek801, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;operating-system#926).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: buildroot-external/package/hardkernel-boot/hardkernel-boot.mk [hunk 1]
```diff
@@ -32,6 +32,63 @@ define HARDKERNEL_BOOT_BUILD_CMDS
...
+	# blx_fix_g12a.sh is appropriate for C4 as well as N2
+	curl -L -o $(@D)/fip/blx_fix.sh https://raw.githubusercontent.com/home-assistant/hassos-blobs/d88ec078470f236d694ba1f9e69a44759ddf41ea/odroid-n2/blx_fix_g12a.sh
+  # get these bits from Neil Armstrong's repo; probably will want to host these
+  # in hassos-blobs (or maybe they're the same as the G12B versions?)
+	curl -L -o $(@D)/fip/acs.bin https://gitlab.com/superna9999/amlogic-boot-fip/-/raw/master/odroid-c4/acs.bin
+	curl -L -o $(@D)/fip/bl301.bin https://gitlab.com/superna9999/amlogic-boot-fip/-/raw/master/odroid-c4/bl301.bin
+
+	bash $(@D)/fip/blx_fix.sh \
+		$(@D)/fip/g12a/bl30.bin $(@D)/fip/zero_tmp $(@D)/fip/bl30_zero.bin \
+		$(@D)/fip/bl301.bin $(@D)/fip/bl301_zero.bin $(@D)/fip/bl30_new.bin \
```
Comment (pvizeli) [inline buildroot-external/package/hardkernel-boot/hardkernel-boot.mk:50]: Make a PR to home-assistant/hassos-blobs for this files

File: Documentation/development.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ Default 2. That is the size of end image in GB.
 
 - intel-nuc
 - odroid-c2
+- odroid-c4 (_experimental_)
```
Comment (pvizeli) [inline Documentation/development.md:39]:
```suggestion
- odroid-c4
```

File: Documentation/kernel.md [hunk 3]
```diff
@@ -13,6 +13,7 @@ Default Kernel tree: 5.4
 | Raspberry Pi 4 | 4.19.127 |
 | Tinker Board | 5.4.63 |
 | Odroid-C2 | 5.7.19 |
+| Odroid-C4 (_experimental_) | 5.9.1 |
```
Comment (pvizeli) [inline Documentation/kernel.md:16]:
```suggestion
| Odroid-C4 | 5.9.1 |
```
Please update also the script for kernel updates in `/scripts`

File: buildroot-external/board/hardkernel/odroid-c4-patches/uboot/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ 	cfg->host_caps = MMC_MODE_8BIT | MMC_MODE_4BIT |
+ 			MMC_MODE_HS_52MHz | MMC_MODE_HS;
+ 	cfg->f_min = DIV_ROUND_UP(SD_EMMC_CLKSRC_24M, CLK_MAX_DIV);
+-	cfg->f_max = 100000000; /* 100 MHz */
++	cfg->f_max = SD_EMMC_CLKSRC_24M;
+ 	cfg->b_max = 511; /* max 512 - 1 blocks */
+ 	cfg->name = dev->name;
+
+--
+2.22.0
```
Comment (pvizeli) [inline buildroot-external/board/hardkernel/odroid-c4-patches/uboot/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch:25]: Move this files into `odroid-c4/patches`

File: buildroot-external/board/hardkernel/odroid-c4/meta [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+BOARD_ID=odroid-n2
+BOARD_NAME="Hardkernel Odroid-C4"
+CHASSIS=embedded
+BOOTLOADER=uboot
+KERNEL_FILE=Image
+BOOT_SYS=mbr
+BOOT_SPL=true
+BOOT_ENV_SIZE=0x2000
+# needs to match something in https://version.home-assistant.io/stable.json
+SUPERVISOR_MACHINE=odroid-n2
```
Comment (pvizeli) [inline buildroot-external/board/hardkernel/odroid-c4/meta:11]: Open a pr on home-assistant/hassio-builder

**Comment by pvizeli** [comment]
> Similar to the Odroid N2, the blobs acs.bin and bl301.bin are provided from a source outside this repo. For now, hardkernel-boot.mk obtains them from Neil Armstrong's repo at https://gitlab.com/superna9999/amlogic-boot. Presumably the "blessed" versions (wherever they come from) would be added to hassos-blobs as with the N2, or maybe the C4's blobs are actually the same as the N2's?

Read the uboot readme for this platform to see how do you need generate this files or update

**Comment by ubergeek801** [comment]
> > Similar to the Odroid N2, the blobs acs.bin and bl301.bin are provided from a source outside this repo.
> > ...
> Read the uboot readme for this platform to see how do you need generate this files or update

I actually have built u-boot and was originally using `acs.bin` and `bl301.bin` from my own build, but I suspected that Home Assistant might be reluctant to accept blobs from a relatively unknown source. I can certainly provide these if acceptable.

File: Documentation/kernel.md (hunk 3 above)
Comment (ubergeek801) [reply Documentation/kernel.md:16]: Using this script updates the kernel version for all Odroid boards; is that what you intended?

Actually the other Odroid configs appear to have been updated to 5.9.1 already by commit 2b0fff31a367da05a8b6e171ab11860d4c7edee0; it's just the documentation files that are out of date.

File: buildroot-external/package/hardkernel-boot/hardkernel-boot.mk (hunk 1 above)
Comment (ubergeek801) [reply buildroot-external/package/hardkernel-boot/hardkernel-boot.mk:50]: Created PR https://github.com/home-assistant/hassos-blobs/pull/1.

File: Documentation/development.md (hunk 2 above)
Comment (ubergeek801) [reply Documentation/development.md]: removed

File: buildroot-external/board/hardkernel/odroid-c4-patches/uboot/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch (hunk 4 above)
Comment (ubergeek801) [reply buildroot-external/board/hardkernel/odroid-c4-patches/uboot/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch]: moved

File: buildroot-external/board/hardkernel/odroid-c4/meta (hunk 5 above)
Comment (ubergeek801) [reply buildroot-external/board/hardkernel/odroid-c4/meta:11]: Created PR https://github.com/home-assistant/builder/pull/58.

**Comment by ubergeek801** [comment]
> I actually have built u-boot and was originally using `acs.bin` and `bl301.bin` from my own build, but I suspected that Home Assistant might be reluctant to accept blobs from a relatively unknown source. I can certainly provide these if acceptable.

In any case, https://github.com/home-assistant/hassos-blobs/pull/1 now takes up this concern.

File: buildroot-external/board/hardkernel/odroid-c4/meta [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# needs to match something in https://version.home-assistant.io/stable.json
+BOARD_ID=odroid-n2
```
Comment (agners) [inline buildroot-external/board/hardkernel/odroid-c4/meta:2]: This should be `odroid-c4` as well

**Review by agners — CHANGES_REQUESTED** [review]
Looks good to me otherwise,

**Comment by pvizeli** [comment]
We need a machine file like: https://github.com/home-assistant/core/blob/dev/machine/odroid-c2

File: buildroot-external/board/hardkernel/odroid-c4/meta (hunk 6 above)
Comment (ubergeek801) [reply buildroot-external/board/hardkernel/odroid-c4/meta:2]: Right; I was hoping we could get the C4 to appear on https://version.home-assistant.io/stable.json (whatever it takes to do that), and then I can update these.

**Comment by ubergeek801** [comment]
> We need a machine file like: https://github.com/home-assistant/core/blob/dev/machine/odroid-c2

Is this something I should submit a PR for? It looks like the C2 and N2 versions are identical, and I don't see why the C4 version couldn't be as well.

**Comment by ubergeek801** [comment]
Working on a PR to `core`. Currently having some trouble with unit tests, which I think has nothing to do with this change, but they need to pass nonetheless...

**Comment by pvizeli** [comment]
Make sure, your branch is rebased and you get also the changes from:
https://github.com/home-assistant/operating-system/pull/938
https://github.com/home-assistant/operating-system/pull/943

**Comment by ubergeek801** [comment]
@pvizeli, I can merge those branches into this one, no problem.

Rebasing is stretching the limits of my git expertise. What do I want to rebase to? Thanks!

**Comment by agners** [comment]
The merge strategy we use is squash and merge, which means we squash all you individual commits into one before merging to the dev branch (see https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits). This does not work with your dev branch currently because you have a non-linear history (your history back-merges dev several times).

I think the easiest approach at this point is to create a new (feature) branch and cherry-pick your commits, e.g.:

```
$ git fetch <remote> (replace remote with whatever points to https://github.com/home-assistant/operating-system/ in `git remote -v`)
$ git checkout -b odroid-c4 <remote>/dev
$ git cherry-pick f9e5f386fcd6289e15e61363c114b2546b73b8ff
$ ....
```

Cherry pick all your non-merge (feature) commits in this list: https://github.com/home-assistant/operating-system/pull/926/commits. You might need to resolve conflicts but they should be rather trivial to fix. Probably best is to create a new pull request with that end result.

**Comment by ubergeek801** [comment]
Ahh yes, squash merging I do understand, and it's been my general preference as well. I will proceed as you suggest with a clean branch with cherry-picked commits merged into it. Thanks!

**Comment by pvizeli** [comment]
Okay, version files updated. Next nightly should build a C4 image to. For me is fine

File: buildroot-external/board/hardkernel/odroid-c4/meta (hunk 6 above)
Comment (agners) [reply buildroot-external/board/hardkernel/odroid-c4/meta:2]: This should now be
```suggestion
BOARD_ID=odroid-c4
```

File: buildroot-external/board/hardkernel/odroid-c4/meta (hunk 5 above)
Comment (agners) [reply buildroot-external/board/hardkernel/odroid-c4/meta:11]: Same here:
```suggestion
SUPERVISOR_MACHINE=odroid-c4
```

**Pull request merged**